### PR TITLE
[Retiarii] end2end

### PR DESCRIPTION
**Pull request by QuanluZhang**
- [x] support LayerChoice and InputChoice
- [x] support new experiment launch approach, i.e., directly launch NAS experiment from python code
- [x] refactor graph ir: merge input/output name to input/output node; graph dump and load
- [ ] improve graph generation and code generation
    - [x] support darts search space
    - [ ] support enas search space
    - [ ] support proxylessnas search space
- [x] support tpe tuner as nas strategy
- [ ] improve debuggability

### CODE REVIEW

File: nni/retiarii/converter/graph_gen.py [hunk 1]
```diff
@@ -243,63 +275,162 @@ def handle_single_node(node):
...
+            edge.head = new_slice_node
+        ir_graph.hidden_nodes.remove(node)
+    
+
+def refine_graph(ir_graph):
+    """
+    Do the following process to simplify graph:
+    1. remove unconnected constant node
+    2. remove unconnected getattr node
+    3. FIXME: merge slice
```
Comment (ultmaster) [inline nni/retiarii/converter/graph_gen.py:368]: Why FIXME?

File: nni/retiarii/converter/graph_gen.py [hunk 2]
```diff
@@ -25,9 +26,11 @@ def _add_edge(ir_graph, node, graph_inputs, node_index, new_node, ignore_first=F
     node_index : Dict
     new_node : Node
         newly created ir node corresponding to `node`
+    output_remap : Dict
```
Comment (ultmaster) [inline nni/retiarii/converter/graph_gen.py:29]: What is output_remap

File: nni/retiarii/converter/graph_gen.py [hunk 3]
```diff
@@ -100,46 +87,60 @@ def create_prim_constant_node(ir_graph, node, module_name):
...
 
-def remove_unconnected_nodes(ir_graph):
+def remove_unconnected_nodes(ir_graph, targeted_type):
     """
     Parameters
     ----------
     ir_graph : Graph
         our ir graph representation
+    targeted_type : str
+        nodes with ```targeted_type``` will be removed from graph is their fanout is 0
```
Comment (ultmaster) [inline nni/retiarii/converter/graph_gen.py:103]: is -> if?

Suggest make ``targeted_type`` optional, or make the function name specific, such as ``remove_unconnected_nodes_with_type``.

File: nni/retiarii/converter/graph_gen.py [hunk 4]
```diff
@@ -243,63 +275,162 @@ def handle_single_node(node):
...
+                                            node_type, attrs)
+            node_index[node] = new_node
         elif node.kind() == 'prim::If':
             last_block_node = handle_if_node(node)
             node_index[node] = last_block_node
-        elif node.kind() == 'prim::GetAttr':
-            pass
+        elif node.kind() == 'prim::Loop':
+            print('mygraph: ', sm_graph)
+            raise RuntimeError('Loop has not been supported yet!')
```
Comment (ultmaster) [inline nni/retiarii/converter/graph_gen.py:310]: What happens to other `prim::` ops?

File: test/convert_test/tpe_strategy.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        chosen = self.cur_sample[str(self.index)]
+        self.index += 1
+        return chosen
+
+class TPEStrategy(BaseStrategy):
+    def __init__(self):
+        self.tpe_sampler = TPESampler()
+        self.model_id = 0
+
+    def run(self, base_model, applied_mutators, trainer):
```
Comment (ultmaster) [inline test/convert_test/tpe_strategy.py:45]: It looks strange that strategy needs an extra trainer. Why don't we merge trainer into base model?

File: nni/retiarii/codegen/pytorch.py [hunk 6]
```diff
@@ -53,8 +53,23 @@ def _format_inputs(node: Node) -> List[str]:
     return inputs
 
 def graph_to_pytorch_model(graph_name: str, graph: Graph) -> str:
-    nodes = graph.nodes
+    def _remove_prefix(names):
```
Comment (liuzhe-lz) [inline nni/retiarii/codegen/pytorch.py:56]: Why is nested function required here?

File: nni/retiarii/converter/op_types.py [hunk 7]
```diff
@@ -1,18 +1,31 @@
 MODULE_EXCEPT_LIST = ['Sequential']
-RETIARII_BASE_OPS = ['Placeholder']
+
 
 class Type:
```
Comment (liuzhe-lz) [inline nni/retiarii/converter/op_types.py:4]: Why use a class here?

File: nni/retiarii/model_apis/inline_mutators.py [hunk 8]
```diff
@@ -0,0 +1,34 @@
+import torch
+import torch.nn as nn
+from typing import (Any, Tuple, List, Optional)
+
+__all__ = ['LayerChoice', 'InputChoice', 'ValueChoice']
+
+class LayerChoice(nn.Module):
+    def __init__(self, candidate_ops: List, label: str = None):
+        super(LayerChoice, self).__init__()
```
Comment (liuzhe-lz) [inline nni/retiarii/model_apis/inline_mutators.py:9]: This is Python 2's style. `super().__init__()` is enough.

File: nni/retiarii/model_apis/nn.py [hunk 9]
```diff
@@ -80,3 +94,7 @@ def __init__(self, *args, **kws):
 ReLU = wrap_module(nn.ReLU)
 Dropout = wrap_module(nn.Dropout)
 Linear = wrap_module(nn.Linear)
+MaxPool2d = wrap_module(nn.MaxPool2d)
```
Comment (liuzhe-lz) [inline nni/retiarii/model_apis/nn.py:97]: Path of this file should contain "pytorch".

File: nni/retiarii/mutator.py [hunk 10]
```diff
@@ -78,9 +78,9 @@ def dry_run(self, model: Model) -> List[List[Choice]]:
         sampler_backup = self.sampler
         recorder = _RecorderSampler()
         self.sampler = recorder
-        self.apply(model)
+        new_model = self.apply(model)
         self.sampler = sampler_backup
-        return recorder.recorded_candidates
+        return recorder.recorded_candidates, new_model
```
Comment (liuzhe-lz) [inline nni/retiarii/mutator.py:83]: Why is dry-run result useful?

File: nni/retiarii/strategy.py [hunk 11]
```diff
@@ -0,0 +1,8 @@
+import abc
+from typing import List
+
+class BaseStrategy(abc.ABC):
+
+    @abc.abstractmethod
+    def run(self, base_model: 'Model', applied_mutators: List['Mutator'], trainer: 'BaseTrainer') -> None:
```
Comment (liuzhe-lz) [inline nni/retiarii/strategy.py:7]: If it only has one method, I don't think there need to be a class.

File: nni/retiarii/trainer/interface.py [hunk 12]
```diff
@@ -17,6 +19,23 @@ class BaseTrainer(abc.ABC):
     Trainer has a ``fit`` function with no return value. Intermediate results and final results should be
     directly sent via ``nni.report_intermediate_result()`` and ``nni.report_final_result()`` functions.
     """
+    def __init__(self, *args, **kwargs):
+        module = self.__class__.__module__
+        if module is None or module == str.__class__.__module__:
```
Comment (liuzhe-lz) [inline nni/retiarii/trainer/interface.py:24]: `module == str.__class__.__module__`?

File: nni/retiarii/converter/graph_gen.py [hunk 13]
```diff
@@ -243,63 +275,162 @@ def handle_single_node(node):
...
+    1. remove unconnected constant node
+    2. remove unconnected getattr node
+    3. FIXME: merge slice
+    """
+    # some constant is not used, for example, function name as prim::Constant
+    remove_unconnected_nodes(ir_graph, targeted_type='prim::Constant')
+    remove_unconnected_nodes(ir_graph, targeted_type='prim::GetAttr')
+    merge_aten_slices(ir_graph)
+
+def _handle_layerchoice(module):
```
Comment (liuzhe-lz) [inline nni/retiarii/converter/graph_gen.py:375]: So you finally decided not to make `LayerChoice` placeholder?

File: nni/retiarii/converter/graph_gen.py [hunk 14]
```diff
@@ -243,63 +275,162 @@ def handle_single_node(node):
...
+    if has_slice_node:
+        assert head_slice_nodes
+    
+    for head_node in head_slice_nodes:
+        slot = 0
+        new_slice_node = ir_graph.add_node(build_full_name(head_node.name, 'merged'), Type.MergedSlice)
+        assert len(head_node.incoming_edges) == 5
+        for edge in head_node.incoming_edges:
+            edge.tail = new_slice_node
+        slot += 5
```
[the remark this answers is not in the document]
Comment (QuanluZhang) [reply nni/retiarii/converter/graph_gen.py:343]: will add comments

File: nni/retiarii/converter/graph_gen.py (hunk 1 above)
Comment (QuanluZhang) [reply nni/retiarii/converter/graph_gen.py:368]: will remove

File: nni/retiarii/converter/op_types.py (hunk 7 above)
Comment (QuanluZhang) [reply nni/retiarii/converter/op_types.py:4]: will change class name and split this class into two classes

**Comment by ultmaster** [comment]
Suggestions:

* Rename `model_apis` to `nn.pytorch`, like [dgl](https://docs.dgl.ai/en/latest/api/python/nn.pytorch.html) did.
* Decorator might be a better idea than tca. An example usage from [mmcv](https://mmcv.readthedocs.io/en/latest/_modules/mmcv/utils/registry.html).

File: nni/retiarii/codegen/pytorch.py (hunk 6 above)
Comment (QuanluZhang) [reply nni/retiarii/codegen/pytorch.py:56]: moved out of this function

File: nni/retiarii/converter/graph_gen.py (hunk 2 above)
Comment (QuanluZhang) [reply nni/retiarii/converter/graph_gen.py:28]: comments have been added below

File: nni/retiarii/converter/graph_gen.py (hunk 3 above)
Comment (QuanluZhang) [reply nni/retiarii/converter/graph_gen.py:102]: updated

File: nni/retiarii/converter/graph_gen.py [hunk 15]
```diff
@@ -207,29 +215,53 @@ def handle_single_node(node):
...
-                assert submodule_name in script_module._modules
-
-                submodule_full_name = build_full_name(module_name, submodule_name)
-                submodule_obj = getattr(module, submodule_name)
-                subgraph, sub_m_attrs = convert_module(script_module._modules[submodule_name],
-                                                       submodule_obj,
-                                                       submodule_full_name, ir_model)
-                # TODO: try not-connected placeholder in TorchScript
+
+                if submodule.inputsAt(0).debugName() == 'self':
```
[the remark this answers is not in the document]
Comment (QuanluZhang) [reply nni/retiarii/converter/graph_gen.py:218]: added

File: nni/retiarii/converter/graph_gen.py [hunk 16]
```diff
@@ -207,29 +215,53 @@ def handle_single_node(node):
...
+                        predecessor = submodule.inputsAt(0).node()
+                        assert predecessor.kind() == 'prim::GetAttr'
+                        assert predecessor.hasAttribute('name')
+                        assert predecessor.inputsAt(0).debugName() == 'self'
+                        predecessor_name = predecessor.s('name')
+                        # FIXME: exchange
+                        submodule_full_name = build_full_name(module_name, [submodule_name, predecessor_name])
+                        predecessor_obj = getattr(module, predecessor_name)
+                        submodule_obj = getattr(predecessor_obj, submodule_name)
+                        print('submodule_name: {}, type of submodule: {}'.format(submodule_name, type(submodule_obj)))
```
[the remark this answers is not in the document]
Comment (QuanluZhang) [reply nni/retiarii/converter/graph_gen.py:241]: removed

File: nni/retiarii/converter/graph_gen.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (QuanluZhang) [reply nni/retiarii/converter/graph_gen.py:342]: added

File: nni/retiarii/converter/graph_gen.py (hunk 13 above)
Comment (QuanluZhang) [reply nni/retiarii/converter/graph_gen.py:374]: let's refactor this part in the future

File: nni/retiarii/converter/op_types.py (hunk 7 above)
Comment (QuanluZhang) [reply nni/retiarii/converter/op_types.py:4]: let me refactor this part in the next pr

File: nni/retiarii/trainer/interface.py [hunk 17]
```diff
@@ -17,6 +19,23 @@ class BaseTrainer(abc.ABC):
     Trainer has a ``fit`` function with no return value. Intermediate results and final results should be
     directly sent via ``nni.report_intermediate_result()`` and ``nni.report_final_result()`` functions.
     """
+    def __init__(self, *args, **kwargs):
+        module = self.__class__.__module__
```
[the remark this answers is not in the document]
Comment (QuanluZhang) [reply nni/retiarii/trainer/interface.py:23]: will refactor it in the next pr

**Comment by QuanluZhang** [comment]
> Suggestions:
> 
> * Rename `model_apis` to `nn.pytorch`, like [dgl](https://docs.dgl.ai/en/latest/api/python/nn.pytorch.html) did.
> * Decorator might be a better idea than tca. An example usage from [mmcv](https://mmcv.readthedocs.io/en/latest/_modules/mmcv/utils/registry.html).

good suggestion, i have finished the first one, and leave the second one in the next pr.

**Review by ultmaster — APPROVED** [review]

**Pull request merged**